lib/fizzy/execute.cpp

```diff
@@ -77,11 +131,61 @@ std::vector<TypeIdx> match_imports(const Module& module,
     }
 
     match_imported_functions(imported_function_types, imported_functions);
+    match_imported_memories(imported_memory_types, imported_memories);
     match_imported_globals(imported_globals_mutability, imported_globals);
 
     return imported_function_types;
 }
 
+std::tuple<bytes_ptr, size_t> allocate_memory(const std::vector<Memory>& module_memories,
+    const std::vector<ImportedMemory>& imported_memories)
+{
+    static const auto bytes_delete = [](bytes* b) noexcept { delete b; };
+    static const auto null_delete = [](bytes*) noexcept {};
+
+    if (module_memories.size() + imported_memories.size() > 1)
+    {
+        // FIXME: turn this into an assert if instantiate is not exposed externally and it only
+        // takes validated modules
+        throw std::runtime_error("Cannot support more than 1 memory section.");
+    }
+    else if (module_memories.size() == 1)
+    {
+        const size_t memory_min = module_memories[0].limits.min;
+        const size_t memory_max =
+            (module_memories[0].limits.max.has_value() ? *module_memories[0].limits.max :
+                                                         MemoryPagesLimit);
+        // FIXME: better error handling
+        if ((memory_min > MemoryPagesLimit) || (memory_max > MemoryPagesLimit))
+            throw std::runtime_error("Cannot exceed hard memory limit of " +
+                                     std::to_string(MemoryPagesLimit * PageSize) + " bytes.");
+
+        // NOTE: fill it with zeroes
+        bytes_ptr memory{new bytes(memory_min * PageSize, 0), bytes_delete};
+        return {std::move(memory), memory_max};
+    }
+    else if (imported_memories.size() == 1)
+    {
+        const size_t memory_min = imported_memories[0].limits.min;
+        const size_t memory_max =
+            (imported_memories[0].limits.max.has_value() ? *imported_memories[0].limits.max :
+                                                           MemoryPagesLimit);
+
+        // FIXME: better error handling
+        if ((memory_min > MemoryPagesLimit) || (memory_max > MemoryPagesLimit))
+            throw std::runtime_error("Imported memory limits cannot exceed hard memory limit of " +
+                                     std::to_string(MemoryPagesLimit * PageSize) + " bytes.");
+
+        bytes_ptr memory{imported_memories[0].data, null_delete};
+        return {std::move(memory), memory_max};
+    }
+    else
+    {
+        bytes_ptr memory{new bytes{}, bytes_delete};
```

Comment:
Apparently this is not correct - no memory should be allocated if none is defined or imported. WebAssembly Studio issues in this case an error like `i32.load requires an imported or defined memory.`
See also for example this test https://github.com/WebAssembly/spec/blob/3b2489d7e570ca0f8d32e9d0b1a1e8255c4212b7/test/core/memory.wast#L22-L25

I would leave it like this here and fix it in another PR.